### PR TITLE
Stringable value objects as router actions

### PR DESCRIPTION
**Pull request by virgiliolino**
**Current situation:** When creating routes, the action must be a string.
With this pull requests we'll be able to pass a value object.

**Reason:** Working  on several DDD projects I try to use, when it may enrich the domain, ValueObjects as an alternative to simple types. This in order to make [implicit concepts more explicit](http://www.markhneedham.com/blog/2009/04/23/ddd-making-implicit-concepts-explicit/).
In my project, action is not a string, its an Action which simply correspond to a value object with the __toString method.

### CODE REVIEW

**Comment by sisve** [comment]
This implementation of Str::isStringable does not support all the scalar values that php can implicitly convert to a string, like integers and floats. I understand that the router does not really want to generate actions to the number 3, but I believe that a public static method on Str should support it.

**Comment by virgiliolino** [comment]
Interesting observation. I'll build some unit tests on the Str::isStringable to develop the concept.
Just, this would introduce, in cases like the router action the need for something like:
```
Str::isStringable($value, $strict = false)
```
So to avoid a router action on number 3 or on value true, in those cases(and avoid headaches on debugging bugs) with a strict set to true, it will just return false. 
I'll think(unit test) on it, and see what happens.

thanks

File: src/Illuminate/Support/Str.php [hunk 1]
```diff
@@ -531,6 +531,17 @@ public static function uuid()
...
 
+    /**
+     * Return true if the param is a string, or an object that maybe treated as string.
+     * @param $value
+     * @return bool
+     */
+    public static function isStringable($value)
+    {
+        return is_string($value) ||
+            (is_object($value) && is_callable([$value, '__toString']));
```
Comment (ntzm) [inline src/Illuminate/Support/Str.php]: > If the target class has __call() magic function implemented, then is_callable will ALWAYS return TRUE for whatever method you call it.

You should use `method_exists` instead

Comment (ntzm) [reply src/Illuminate/Support/Str.php]: Also, I believe the way to catch all the possible stringable values is:

`return $value === null || is_scalar($value) || method_exists($value, '__toString');`

**Comment by taylorotwell** [comment]
No plans to add this. Not sure we're adding real value here.

**Pull request closed without merging**